proposal for splitting the Nix store (#3)
nix-community/projects: proposals/store.md

There exists a [security team](https://nixos.org/community/teams/security.html) and an [infrastructure team](https://nixos.org/community/teams/infrastructure.html) tending to critical assets.
[Security-relevant issues and pull-requests](https://github.com/NixOS/nixpkgs/labels/1.severity%3A%20security) are automatically tagged by a [vulnerability tracking system](https://github.com/nix-community/vulnix/).

Different components have different ownership structures.
For Nix itself, only maintainers have write access, and only the principal author can publish releases.
Nixpkgs and NixOS have a much broader committer community with largely implicit rules.
There is a high degree of social trust involved, and active developers and maintainers, while culturally and geographically diverse, usually have ongoing working relationships marked by diligent reviews and critical discussions.

## How will you address the challenge described? Give an overview of your work and why it is significant, specifically addressing the challenge.

> Explain what the field will learn from your proposed work and how it contributes to the long-term sustainability of the technology.

Reviewer: This sentence is probably a leftover note?

> ## How will you address the challenge described? Give an overview of your work and why it is significant, specifically addressing the challenge.
> 
> Explain what the field will learn from your proposed work and how it contributes to the long-term sustainability of the technology.

Author: ```suggestion
## How will you address the challenge described? Give an overview of your work and why it is significant, specifically addressing the challenge. Explain what the field will learn from your proposed work and how it contributes to the long-term sustainability of the technology.
```

@iFreilicht Thanks for catching that. It's an artifact of splitting lines by sentence for easier reviews and suggestions.

### How our work will contribute to sustainability of Nix, the Nix ecosystem, and FOSS more generally

Opening up the Nix store specification for alternative implementations will foster competition for code quality and performance, make the ecosystem less reliant on any one implementation, and drive innovation while incentivising keeping existing interfaces stable.
Both aspects are major concerns for Nix users: Nix is well-known for its stability guarantees but also has some notorious performance bottlenecks.

Well-defined interfaces, accompanying documentation, and a more comprehensive test suite will ease learning and adoption for developers, making quality contributions more likely.
That would reduce maintainers’ work load and contribute to making the maintainer role more appealing; having a separate component as a domain of ownership would also help with managing responsibilities and expectations in the long run.

> 

Reviewer 2: Maybe let's add something like

```suggestion

Eventually, as the Nix store is adopted by other projects, we will be able to make use of them directly in Nixpkgs, for example through [RFC 92](https://github.com/NixOS/rfcs/blob/master/rfcs/0092-plan-dynamism.md).
This can close the loop between system and language-specific package managers, removing their current disconnect and enabling a virtuous cycle of collaboration between them.

```

I am a member of the [Nix maintainers team](https://discourse.nixos.org/t/nix-team-creation/22228).
This project proposal was developed and written collaboratively by Nix maintainers and Tvix developers.
It aims to implement [RFC 134](https://github.com/nixos/rfcs/blob/master/rfcs/0134-nix-store-layer.md), which met broad consensus as crucial for maintaining and developing the Nix ecosystem at scale, but currently has no perspective of being implemented due to lack of resources we can allocate to sustainability measures such as testing, refactoring, and contributor-oriented documentation.
Maintainers only have capacity to review and shepherd contributions following settled design decisions, and are committed to that task as is evidenced by our [meeting minutes](https://discourse.nixos.org/search?q=Nix%20team%20meeting%20minutes%20%23%20%23dev%3Anix%20order%3Alatest_topic) and our [first half-yearly report](https://discourse.nixos.org/t/nix-team-report-2022-10-2023-03/27486).

> We have explicit support by Cabal maintainers to add the Nix store backend, and the Nix community has a lot of ties and overlap with the Haskell community.

Reviewer 2: ```suggestion
We have explicit support by [the Cabal team](https://github.com/haskell/cabal/issues/9089) to add the Nix store backend, we have a Cabal maintainer among the contributors, and the Nix community has a lot of ties and overlap with the Haskell community.
```

edit: On the other hand the discussion there may take a while to settle
